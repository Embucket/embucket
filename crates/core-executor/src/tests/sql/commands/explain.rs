use crate::test_query;

const SETUP_QUERY: &str = r"CREATE OR REPLACE TABLE testing (name VARCHAR)
AS SELECT * FROM VALUES
  ('yes'),
  ('no')";

test_query!(
    explain_visitors_basic,
    "EXPLAIN SELECT dateadd(day, 30, '20-12-2022');",
    snapshot_path = "explain"
);

test_query!(
    explain_analyze_visitors_basic,
    "EXPLAIN ANALYZE SELECT dateadd(day, 30, '20-12-2022');",
    snapshot_path = "explain"
);

test_query!(
    explain_verbose_visitors_basic,
    "EXPLAIN VERBOSE SELECT dateadd(day, 30, '20-12-2022');",
    snapshot_path = "explain"
);

test_query!(
    explain_references_basic,
    "EXPLAIN SELECT name FROM testing;",
    setup_queries = [SETUP_QUERY],
    snapshot_path = "explain"
);

test_query!(
    explain_analyze_references_basic,
    "EXPLAIN ANALYZE SELECT name FROM testing;",
    setup_queries = [SETUP_QUERY],
    snapshot_path = "explain"
);

test_query!(
    explain_verbose_references_basic,
    "EXPLAIN VERBOSE SELECT name FROM testing;",
    setup_queries = [SETUP_QUERY],
    snapshot_path = "explain"
);
