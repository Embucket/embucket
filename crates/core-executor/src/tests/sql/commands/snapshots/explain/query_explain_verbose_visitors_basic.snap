---
source: crates/core-executor/src/tests/sql/commands/explain.rs
description: "\"EXPLAIN VERBOSE SELECT dateadd(day, 30, '20-12-2022');\""
---
Ok(
    [
        "+------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+",
        "| plan_type                                                  | plan                                                                                                                                                                                    |",
        "+------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+",
        "| initial_logical_plan                                       | Projection: dateadd(Utf8(\"day\"), Int64(30), Utf8(\"20-12-2022\"))                                                                                                                         |",
        "|                                                            |   EmptyRelation: rows=1                                                                                                                                                                 |",
        "| logical_plan after apply_function_rewrites                 | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after LikeTypeAnalyzer                        | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after custom_type_coercion                    | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after resolve_grouping_function               | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after type_coercion                           | Projection: dateadd(Utf8(\"day\"), Int64(30), CAST(Utf8(\"20-12-2022\") AS Timestamp(Nanosecond, None)))                                                                                    |",
        "|                                                            |   EmptyRelation: rows=1                                                                                                                                                                 |",
        "| logical_plan after IcebergTypesAnalyzer                    | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after CastAnalyzer                            | Projection: dateadd(Utf8(\"day\"), Int64(30), to_timestamp(Utf8(\"20-12-2022\"))) AS dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))                                                      |",
        "|                                                            |   EmptyRelation: rows=1                                                                                                                                                                 |",
        "| logical_plan after UnionSchemaAnalyzer                     | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| analyzed_logical_plan                                      | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_nested_union                  | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after simplify_expressions                    | Projection: TimestampNanosecond(1674086400000000000, None) AS dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))                                                                         |",
        "|                                                            |   EmptyRelation: rows=1                                                                                                                                                                 |",
        "| logical_plan after replace_distinct_aggregate              | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_join                          | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after decorrelate_predicate_subquery          | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after scalar_subquery_to_join                 | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after decorrelate_lateral_join                | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after extract_equijoin_predicate              | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_duplicated_expr               | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_filter                        | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_cross_join                    | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_limit                         | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after propagate_empty_relation                | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_one_union                     | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after filter_null_join_keys                   | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_outer_join                    | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after push_down_limit                         | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after push_down_filter                        | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after single_distinct_aggregation_to_group_by | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_group_by_constant             | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after common_sub_expression_eliminate         | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after optimize_projections                    | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after split_ordered_aggregates                | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_nested_union                  | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after simplify_expressions                    | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after replace_distinct_aggregate              | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_join                          | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after decorrelate_predicate_subquery          | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after scalar_subquery_to_join                 | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after decorrelate_lateral_join                | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after extract_equijoin_predicate              | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_duplicated_expr               | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_filter                        | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_cross_join                    | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_limit                         | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after propagate_empty_relation                | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_one_union                     | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after filter_null_join_keys                   | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_outer_join                    | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after push_down_limit                         | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after push_down_filter                        | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after single_distinct_aggregation_to_group_by | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after eliminate_group_by_constant             | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after common_sub_expression_eliminate         | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after optimize_projections                    | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan after split_ordered_aggregates                | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| logical_plan                                               | Projection: TimestampNanosecond(1674086400000000000, None) AS dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))                                                                         |",
        "|                                                            |   EmptyRelation: rows=1                                                                                                                                                                 |",
        "| initial_physical_plan                                      | ProjectionExec: expr=[1674086400000000000 as dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))]                                                                                         |",
        "|                                                            |   PlaceholderRowExec                                                                                                                                                                    |",
        "|                                                            |                                                                                                                                                                                         |",
        "| initial_physical_plan_with_stats                           | ProjectionExec: expr=[1674086400000000000 as dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))], statistics=[Rows=Exact(1), Bytes=Exact(8), [(Col[0]:)]]                                |",
        "|                                                            |   PlaceholderRowExec, statistics=[Rows=Exact(1), Bytes=Exact(0), []]                                                                                                                    |",
        "|                                                            |                                                                                                                                                                                         |",
        "| initial_physical_plan_with_schema                          | ProjectionExec: expr=[1674086400000000000 as dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))], schema=[dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\")):Timestamp(Nanosecond, None)] |",
        "|                                                            |   PlaceholderRowExec, schema=[]                                                                                                                                                         |",
        "|                                                            |                                                                                                                                                                                         |",
        "| physical_plan after EliminateEmptyDataSourceExec           | ProjectionExec: expr=[1674086400000000000 as dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))]                                                                                         |",
        "|                                                            |   PlaceholderRowExec                                                                                                                                                                    |",
        "|                                                            |                                                                                                                                                                                         |",
        "| physical_plan after RemoveExecAboveEmpty                   | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after OutputRequirements                     | OutputRequirementExec: order_by=[], dist_by=Unspecified                                                                                                                                 |",
        "|                                                            |   ProjectionExec: expr=[1674086400000000000 as dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))]                                                                                       |",
        "|                                                            |     PlaceholderRowExec                                                                                                                                                                  |",
        "|                                                            |                                                                                                                                                                                         |",
        "| physical_plan after aggregate_statistics                   | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after join_selection                         | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after LimitedDistinctAggregation             | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after FilterPushdown                         | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after EnforceDistribution                    | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after CombinePartialFinalAggregate           | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after EnforceSorting                         | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after OptimizeAggregateOrder                 | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after ProjectionPushdown                     | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after coalesce_batches                       | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after coalesce_async_exec_input              | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after OutputRequirements                     | ProjectionExec: expr=[1674086400000000000 as dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))]                                                                                         |",
        "|                                                            |   PlaceholderRowExec                                                                                                                                                                    |",
        "|                                                            |                                                                                                                                                                                         |",
        "| physical_plan after LimitAggregation                       | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after LimitPushPastWindows                   | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after LimitPushdown                          | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after ProjectionPushdown                     | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after EnsureCooperative                      | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after FilterPushdown(Post)                   | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan after SanityCheckPlan                        | SAME TEXT AS ABOVE                                                                                                                                                                      |",
        "| physical_plan                                              | ProjectionExec: expr=[1674086400000000000 as dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))]                                                                                         |",
        "|                                                            |   PlaceholderRowExec                                                                                                                                                                    |",
        "|                                                            |                                                                                                                                                                                         |",
        "| physical_plan_with_stats                                   | ProjectionExec: expr=[1674086400000000000 as dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))], statistics=[Rows=Exact(1), Bytes=Exact(8), [(Col[0]:)]]                                |",
        "|                                                            |   PlaceholderRowExec, statistics=[Rows=Exact(1), Bytes=Exact(0), []]                                                                                                                    |",
        "|                                                            |                                                                                                                                                                                         |",
        "| physical_plan_with_schema                                  | ProjectionExec: expr=[1674086400000000000 as dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\"))], schema=[dateadd(Utf8(\"day\"),Int64(30),Utf8(\"20-12-2022\")):Timestamp(Nanosecond, None)] |",
        "|                                                            |   PlaceholderRowExec, schema=[]                                                                                                                                                         |",
        "|                                                            |                                                                                                                                                                                         |",
        "+------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+",
    ],
)
